### PR TITLE
[MOB-2605] persistent read state

### PR DESCRIPTION
**Pull request by roninopf**
## 🔹 JIRA Ticket(s) if any

* [MOB-2605](https://iterable.atlassian.net/browse/MOB-2605)

## ✏️ Description

* skips over showing in-app messages that have already been read

### CODE REVIEW

File: swift-sdk/Internal/InAppManager+Functions.swift [hunk 1]
```diff
@@ -52,25 +52,26 @@ struct MessagesProcessor {
...
+        } else {
+            ITBDebug("delegate returned skip")
+            return .skip(message)
+        }
     }
     
     private func getFirstProcessableTriggeredMessage() -> IterableInAppMessage? {
         messagesMap.values
             .filter(MessagesProcessor.isProcessableTriggeredMessage)
+            .filter({ !$0.read })
```
Comment (tapashmajumder) [inline swift-sdk/Internal/InAppManager+Functions.swift:74]: You can add this to `MessagesProcessor.isProcessableTriggeredMessage` method. Not need to do filtering twice.

File: swift-sdk/Internal/InAppManager+Functions.swift [hunk 2]
```diff
@@ -52,25 +52,26 @@ struct MessagesProcessor {
...
-            ITBDebug("isOkToShowNow")
-            if inAppDelegate.onNew(message: message) == .show {
-                ITBDebug("delegate returned show")
-                return .show(message)
-            } else {
-                ITBDebug("delegate returned skip")
-                return .skip(message)
-            }
-        } else {
+        guard inAppDisplayChecker.isOkToShowNow(message: message) else {
```
Comment (tapashmajumder) [inline swift-sdk/Internal/InAppManager+Functions.swift:55]: +1

File: swift-sdk/Internal/InAppManager+Functions.swift [hunk 3]
```diff
@@ -122,20 +122,34 @@ struct MessagesObtainedHandler {
...
         
+        var messagesOverwritten = 0
         var newMessagesMap = OrderedDictionary<String, IterableInAppMessage>()
-        messages.forEach {
-            if let existingMessage = messagesMap[$0.messageId] {
-                newMessagesMap[$0.messageId] = existingMessage
+        messages.forEach { serverMessage in
+            let messageId = serverMessage.messageId
+            if let existingMessage = messagesMap[messageId] {
+                if Self.shouldOverwrite(clientMessage: existingMessage, withServerMessage: serverMessage) {
```
Comment (tapashmajumder) [inline swift-sdk/Internal/InAppManager+Functions.swift:130]: Have we added tests for this?

File: swift-sdk/Internal/InAppManager+Functions.swift [hunk 4]
```diff
@@ -122,20 +122,34 @@ struct MessagesObtainedHandler {
...
+                    newMessagesMap[messageId] = existingMessage
+                }
             } else {
-                newMessagesMap[$0.messageId] = $0
+                newMessagesMap[messageId] = serverMessage
             }
         }
         
-        return MergeMessagesResult(inboxChanged: removedInboxCount + addedInboxCount > 0,
+        return MergeMessagesResult(inboxChanged: removedInboxCount + addedInboxCount + messagesOverwritten > 0,
```
Comment (tapashmajumder) [inline swift-sdk/Internal/InAppManager+Functions.swift:141]: We should also add a test for this.

File: tests/swift-sdk-swift-tests/in-app-tests/InAppMessageProcessorTests.swift [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+        let localMessage = makeEmptyInboxMessage(messageId)
+        
+        let serverMessage = makeEmptyInboxMessage(messageId)
+        serverMessage.read = true
+        
+        let messagesMap: OrderedDictionary<String, IterableInAppMessage> = [messageId: localMessage]
+        let newMessages = [serverMessage]
+        
+        let result = MessagesObtainedHandler(messagesMap: messagesMap,
+                                              messages: newMessages).handle()
```
Comment (tapashmajumder) [inline tests/swift-sdk-swift-tests/in-app-tests/InAppMessageProcessorTests.swift:22]: nit: formatting. `messagesMap` and `messages` should align.

File: tests/swift-sdk-swift-tests/in-app-tests/InAppPersistenceTests.swift [hunk 6]
```diff
@@ -57,4 +57,70 @@ class InAppPersistenceTests: XCTestCase {
...
+                    XCTFail("could not get in-app message for test")
+                    return
+                }
+                
+                XCTAssertTrue(firstMessage.read)
+        }
+    }
+    
+    func testPersistentReadStateForMultipleUsers() {
+//        let secondUser = "another@email.com"
```
Comment (tapashmajumder) [inline tests/swift-sdk-swift-tests/in-app-tests/InAppPersistenceTests.swift:104]: Please add more detail or remove this test.

File: tests/swift-sdk-swift-tests/in-app-tests/InAppPersistenceTests.swift [hunk 7]
```diff
@@ -57,4 +57,70 @@ class InAppPersistenceTests: XCTestCase {
...
+        
+        
+    }
+    
+    private static let email = "user@example.com"
+    
+    private let emptyInAppContent = IterableHtmlInAppContent(edgeInsets: .zero, html: "")
```
Comment (tapashmajumder) [inline tests/swift-sdk-swift-tests/in-app-tests/InAppPersistenceTests.swift:111]: nit: Please make static.

File: tests/swift-sdk-swift-tests/in-app-tests/InAppPersistenceTests.swift [hunk 8]
```diff
@@ -57,4 +57,70 @@ class InAppPersistenceTests: XCTestCase {
...
+    
+    private func getInboxMessage(_ id: String = "", _ read: Bool) -> IterableInAppMessage {
```
Comment (tapashmajumder) [inline tests/swift-sdk-swift-tests/in-app-tests/InAppPersistenceTests.swift:113]: nit: mark as static

File: tests/swift-sdk-swift-tests/in-app-tests/InAppPersistenceTests.swift [hunk 9]
```diff
@@ -57,4 +57,70 @@ class InAppPersistenceTests: XCTestCase {
             XCTFail("conversion to string failed")
         }
     }
+    
+    func testPersistentReadStateAfterLogout() {
```
Comment (tapashmajumder) [inline tests/swift-sdk-swift-tests/in-app-tests/InAppPersistenceTests.swift:61]: This test is actually not doing what it is supposed to. We need to make sure that server state is reflected on logout and re-login. Here the server state is `read = false`. 

The test is passing for wrong reasons. Please remove this test.

File: tests/swift-sdk-swift-tests/in-app-tests/InAppPersistenceTests.swift [hunk 10]
```diff
@@ -57,4 +57,70 @@ class InAppPersistenceTests: XCTestCase {
...
+        inboxModel.set(read: true, forMessage: inboxModel.message(atIndexPath: IndexPath(row: 0, section: 0)))
+        
+        internalAPI.email = nil
+        
+        internalAPI.email = InAppPersistenceTests.email
+        
+        XCTAssertTrue(firstMsg.read)
+    }
+    
+    func testPersistentReadStateFromServerPayload() {
```
Comment (tapashmajumder) [inline tests/swift-sdk-swift-tests/in-app-tests/InAppPersistenceTests.swift:85]: This test should initialize with read = false initially, Then have messages arrive from server with read = true and validate that indeed read is true.

File: tests/swift-sdk-swift-tests/in-app-tests/InAppPersistenceTests.swift [hunk 11]
```diff
@@ -57,4 +57,70 @@ class InAppPersistenceTests: XCTestCase {
...
+        let messages = [getInboxMessage("", true)]
```
Comment (tapashmajumder) [inline tests/swift-sdk-swift-tests/in-app-tests/InAppPersistenceTests.swift:86]: set read = true here.

File: tests/swift-sdk-swift-tests/in-app-tests/InAppPersistenceTests.swift (hunk 10 above)
Comment (tapashmajumder) [reply tests/swift-sdk-swift-tests/in-app-tests/InAppPersistenceTests.swift:85]: please see comments below

File: tests/swift-sdk-swift-tests/in-app-tests/InAppPersistenceTests.swift [hunk 12]
```diff
@@ -57,4 +57,70 @@ class InAppPersistenceTests: XCTestCase {
...
+        
+        let mockInAppFetcher = MockInAppFetcher()
+        
+        let internalAPI = IterableAPIInternal.initializeForTesting(inAppFetcher: mockInAppFetcher)
+        
+        mockInAppFetcher.mockMessagesAvailableFromServer(internalApi: internalAPI, messages: messages)
```
Comment (tapashmajumder) [inline tests/swift-sdk-swift-tests/in-app-tests/InAppPersistenceTests.swift:92]: set read = true here.

File: swift-sdk/Internal/InAppManager+Functions.swift (hunk 3 above)
Comment (roninopf) [reply swift-sdk/Internal/InAppManager+Functions.swift:130]: It's now covered with the test in `InAppMessageProcessorTests`

File: swift-sdk/Internal/InAppManager+Functions.swift (hunk 4 above)
Comment (roninopf) [reply swift-sdk/Internal/InAppManager+Functions.swift:141]: It's now covered with the test in `InAppMessageProcessorTests`

**Review by tapashmajumder — APPROVED** [review]
+1

**Pull request merged**